upload-server/pkg/redislocker/redislocker.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/go-redsync/redsync/v4/redis/goredis/v9"
+	"github.com/redis/go-redis/v9"
+	"github.com/tus/tusd/v2/pkg/handler"
+)
+
+var (
+	LockExchangeChannelPrefix = "tusd_lock_release_request_%s"
+	RetryInterval             = 500 * time.Millisecond
+	LockExpiry                = 8 * time.Second
+)
+
+const (
+	RELEASE_REQUEST = "release requested"
+)
+
+type LockerOption func(l *RedisLocker)
+
+func WithLogger(logger *slog.Logger) LockerOption {
+	return func(l *RedisLocker) {
+		l.logger = logger
+	}
+}
+
+func New(uri string, lockerOptions ...LockerOption) (*RedisLocker, error) {
+	connection, err := redis.ParseURL(uri)
+	if err != nil {
+		return nil, err
+	}
+	client := redis.NewClient(connection)
+	rs := redsync.New(goredis.NewPool(client))
+
+	locker := &RedisLocker{
+		rs:    rs,
+		redis: client,
+	}
+	for _, option := range lockerOptions {
+		option(locker)
+	}
+	//defaults
+	if locker.logger == nil {
+		locker.logger = slog.Default()
+	}
+
+	return locker, nil
+}
+
+type LockExchange interface {
+	Listen(ctx context.Context, id string, callback func())
+	Request(ctx context.Context, id string)
+}
+
+type RedisLockExchange struct {
+	client *redis.Client
+}
+
+func (e *RedisLockExchange) channelName(id string) string {
+	return fmt.Sprintf(LockExchangeChannelPrefix, id)
+}
+
+func (e *RedisLockExchange) Listen(ctx context.Context, id string, callback func()) {
+	psub := e.client.PSubscribe(ctx, e.channelName(id))
+	c := psub.Channel()
+	select {
+	case <-c:
+		callback()
+	case <-ctx.Done():
+		return
+	}
+}
+
+func (e *RedisLockExchange) Request(ctx context.Context, id string) {
+
+	e.client.Publish(ctx, e.channelName(id), RELEASE_REQUEST)
+}
+
+type RedisLocker struct {
+	rs     *redsync.Redsync
+	redis  *redis.Client
+	logger *slog.Logger
+}
+
+func (locker *RedisLocker) UseIn(composer *handler.StoreComposer) {
+	composer.UseLocker(locker)
+}
+
+func (locker *RedisLocker) NewLock(id string) (handler.Lock, error) {
+	mutex := locker.rs.NewMutex(id, redsync.WithExpiry(LockExpiry))
+	return &redisLock{
+		id:    id,
+		mutex: mutex,
+		exchange: &RedisLockExchange{
+			client: locker.redis,
+		},
+		logger: locker.logger.With("upload_id", id),
+	}, nil
+}
+
+type redisLock struct {
+	id       string
+	mutex    *redsync.Mutex
+	ctx      context.Context
+	cancel   func()
+	exchange LockExchange
+	logger   *slog.Logger
+}
+
+func (l *redisLock) Lock(ctx context.Context, releaseRequested func()) error {
+	if err := l.aquireLock(ctx); err != nil {
+		if err := l.retryLock(ctx); err != nil {
+			return err
+		}
+	}
+	go l.exchange.Listen(l.ctx, l.id, releaseRequested)
+	go func() {
+		if err := l.keepAlive(l.ctx); err != nil {
+			l.cancel()
+			if releaseRequested != nil {
+				releaseRequested()
+			}
+		}
+	}()
+	return nil
+}
+
+func (l *redisLock) aquireLock(ctx context.Context) error {
+	if err := l.mutex.TryLockContext(ctx); err != nil {
+		return err
+	}
+
+	l.ctx, l.cancel = context.WithCancel(context.Background())
```

Comment:
```suggestion
	l.ctx, l.cancel = context.WithCancel(ctx)
```

Maybe it would make sense to bind `l.ctx` to `ctx`, so if `ctx` is cancelled `l.ctx` is cancelled as well, making sure that `l.exchange.Listen` and `l.keepAlive` are stopped.

Reply:
I assumed this context was centered around acquiring the lock, so it might timeout or be canceled too quickly? As opposed to a context that lives for the full lifecycle of an acquired lock.

Reply:
I double checked and you are correct. `ctx`'s lifetime is for acquiring the lock. Please ignore my comment above.